instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py

```diff
@@ -176,46 +184,63 @@ def _patched_api_call(self, original_func, instance, args, kwargs):
 
-            if span.is_recording():
-                if "ResponseMetadata" in result:
-                    metadata = result["ResponseMetadata"]
-                    req_id = None
-                    if "RequestId" in metadata:
-                        req_id = metadata["RequestId"]
-                    elif "HTTPHeaders" in metadata:
-                        headers = metadata["HTTPHeaders"]
-                        if "x-amzn-RequestId" in headers:
-                            req_id = headers["x-amzn-RequestId"]
-                        elif "x-amz-request-id" in headers:
-                            req_id = headers["x-amz-request-id"]
-                        elif "x-amz-id-2" in headers:
-                            req_id = headers["x-amz-id-2"]
-
-                    if req_id:
-                        span.set_attribute(
-                            "aws.request_id", req_id,
-                        )
-
-                    if "RetryAttempts" in metadata:
-                        span.set_attribute(
-                            "retry_attempts", metadata["RetryAttempts"],
-                        )
-
-                    if "HTTPStatusCode" in metadata:
-                        span.set_attribute(
-                            SpanAttributes.HTTP_STATUS_CODE,
-                            metadata["HTTPStatusCode"],
-                        )
-
-            if error:
-                raise error
+def _apply_response_attributes(span: Span, result):
+    if result is None or not span.is_recording():
+        return
 
-            return result
+    metadata = result.get("ResponseMetadata")
+    if metadata is None:
+        return
+
+    request_id = metadata.get("RequestId")
+    if request_id is None:
+        headers = metadata.get("HTTPHeaders")
+        if headers is not None:
+            request_id = (
+                headers.get("x-amzn-RequestId")
+                or headers.get("x-amz-request-id")
+                or headers.get("x-amz-id-2")
+            )
+    if request_id:
+        # TODO: update when semantic conventions exist
+        span.set_attribute("aws.request_id", request_id)
+
+    retry_attempts = metadata.get("RetryAttempts")
+    if retry_attempts is not None:
+        # TODO: update when semantic conventinos exists
+        span.set_attribute("retry_attempts", retry_attempts)
+
+    status_code = metadata.get("HTTPStatusCode")
+    if status_code is not None:
+        span.set_attribute(SpanAttributes.HTTP_STATUS_CODE, status_code)
+
+
+def _determine_call_context(
+    client: BaseClient, args
+) -> Optional[_AwsSdkCallContext]:
+    try:
+        operation = args[0]
+        params = args[1]
+        call_context = _AwsSdkCallContext(client, operation, params)
+
+        logger.debug(
+            "AWS SDK invocation: %s %s",
+            call_context.service,
+            call_context.operation,
+        )
+
+        return call_context
+    except Exception as ex:  # pylint:disable=broad-except
```

Comment:
this is absolutely required? would be better if we caught specific exceptions here instead of everything.

Reply:
since this is trying to extract parameters from the internals of the botocore lib there could be unforeseen errors, e.g. when internals of botocore change, that wouldn't be handled here. This in turn would affect existing user code which is against one of the design principles of opentelemetry. So that is actually the reason for the broad except here.

Reply:
well, i guess is certainly possible to use `getattr` and `hasattr` to safely access attributes but the question is then what to do if one of the attributes is actually missing. From what i can see there are 2 possibilities. 1) assign a default value or use None, or 2) throw a (typed) exception.
Option 1) is probably not a good idea since it will either produce incorrect data or fail at a later point in time when a `None` field is accessed by an extension.
Option 2) will clutter the code for extracting the attributes since each attribute needs to be checked if it exists and throw an exception if not. Additionally, i don't think that there is much benefit in this approach over broadly catching a possible exception. Since the try/catch block only covers the creation of `_AWSCallContext` there is imo no harm in catching the general exception as for creating the span an instance of `_AWSCallContext` is required in any case.

Reply:
IMO

- if something is not available expectedly (may or may not be set by design), it should be ignored and not recorded as an attribute.
- if something is not available unexpectedly (internal structure changed), it should again be ignored and not recorded as an attribute but in this case an additional warning should be logged so the issue surfaces up and we become aware of it.

Reply:
>  i don't think that there is much benefit in this approach over broadly catching a possible exception. Since the try/catch block only covers the creation of _AWSCallContext there is imo no harm in catching the general exception as for creating the span an instance of _AWSCallContext is required in any case.

One big downside is that if one attribute fails to be extracted, we'll not record anything. If we do it defensively per attribute, we'll still record what we can extract and ignore what we can't. IMO in the end, some telemetry is better than no telemetry for the end-user.

Reply:
Updated the PR and most of the attributes in `_AWSCallContext` are now extracted defensively. Attributes `service` and `operation` are extracted the normal way though since instrumentation without them does not really make much sense.
I kept the broad exception as safety net to avoid breaking user code (e.g. if an extracted attribute is actually a property that might raise some unforeseen exception).